### PR TITLE
[Website]: Adopting Apache Arrow at CloudQuery

### PR DESCRIPTION
**Pull request by yevgenypats**
Hi Team, we wrote a case study about our journey and we decided to adopt arrow at [CloudQuery](https://github.com/cloudquery/cloudquery). Once reviewed and merged we would also like to cross post it on our website - similar to [duckdb post](https://arrow.apache.org/blog/2021/12/03/arrow-duckdb/).

Thank you! cc @zeroshade

### CODE REVIEW

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (zeroshade) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:53]:
```suggestion
This is where Arrow comes in. Apache Arrow defines a language-independent columnar format for flat and hierarchical data, and brings the following advantages:
```

**Review by zeroshade — COMMENTED** [review]
I think this is a great contribution. I've added a few suggestions to the writing and if you have any concrete numbers and tests you've run that you are able to share that would definitely improve this a bit. People always love cool performance diagrams that show the precise benefits :smile:

**Comment by yevgenypats** [comment]
Thanks for the review! I've added also a cross-post link.

We don't have interesting performance benchmarks yet (might add another blog in the future focused on performance. We did some basic performance testing just to make sure it performance at least in same ballpark as what we had before).

For us really the most important part was as mentioned in the blog is less the performance aspect (as long as it doesn't make performance worse then what we had with our in-house implementation of-course) and more 1) the language independent design as we are interested to expand [CloudQuery SDK ](https://github.com/cloudquery/plugin-sdk/) to other languages in the future and 2) The rich type system with nested types support and 3) that is already quite adopted by different solutions which makes it easier for us as an ELT framework.

**Comment by yevgenypats** [comment]
@zeroshade thx for the review! Can you please take another look at this one (applied all the fixes)?

**Review by zeroshade — APPROVED** [review]
LGTM

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (ianmcook) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:51]: ItI think it is preferable to add this PNG file into the PR at the path `img/20230424-xkcd-standards.png` and change the `src` here to point to that.

Comment (ianmcook) [reply _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:51]: Actually xkcd says [hotlinking is fine](https://xkcd.com/license.html) but they prefer for folks include a link to the original comic (https://xkcd.com/927/).

Comment (yevgenypats) [reply _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:51]: Yeah that's why I just pointed to the original hotlink. Should I change it?

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (ianmcook) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:48]: This resolves my below comment
```suggestion
Before Arrow, we used our own type system that supported more than 14 types. This served us well, but we started to hit limitations in various use-cases. For example, in database to database replication, we needed to support many more types, including nested types. Also, performance-wise, lots of the time spent in an ELT process is around converting data from one format to another, so we wanted to take a step back and see if we can avoid this [famous XKCD](https://xkcd.com/927/) (by building yet another format):
```

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (ianmcook) [reply _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:51]: Cool, thanks. I think we should just hyperlink to https://xkcd.com/927/ as I suggested above.

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (ianmcook) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:58]:
```suggestion
2. Performance: Arrow adoption is rising especially in columnar based databases ([DuckDB](https://duckdb.org/2021/12/03/duck-arrow.html), [ClickHouse](https://clickhouse.com/docs/en/integrations/data-formats/arrow-avro-orc), [BigQuery](https://cloud.google.com/bigquery/docs/samples/bigquerystorage-arrow-quickstart)) and file formats ([Parquet](https://arrow.apache.org/docs/python/parquet.html)) which makes it easier to write CloudQuery destination or source plugins for databases that already support Arrow as well as much more efficient as we remove the need for additional serialization and transformation step. Moreover, just the performance of sending Arrow format from source plugin to destination is already more performant and memory efficient, given its “zero-copy” nature and not needing serialization/deserialization. 
```

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (ianmcook) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:63]:
```suggestion
Adopting Apache Arrow as the CloudQuery in-memory type system enables us to gain better performance, data interoperability and developer experience. Some plugins that are going to gain an immediate boost of rich type systems are our database-to-database replication plugins such as [PostgreSQL CDC](https://www.cloudquery.io/docs/plugins/sources/postgresql/overview) source plugin (and all [database destinations](https://www.cloudquery.io/docs/plugins/destinations/overview)) that are going to get support for all available types including nested ones.
```

**Review by ianmcook — APPROVED** [review]
I suggested a couple very small changes. Otherwise LGTM. Thanks for this!

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (wjones127) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:58]: It's a little unclear to me: do you use Arrow arrays at all, or just the types? And if it's just the data types, them maybe mentioning that Arrow uses flatbuffers to define its type system which is part of why exchanging them is performant.

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (wjones127) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:42]: So, if I understand correctly, another way of saying this is:

> Sources and destinations each have their own way to represent the schema of some dataset. In order to avoid having to write a conversion between every pair of source and destination (which would be difficult to maintain), we choose an intermediate schema representation that all sources and destinations can be converted to and from.

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (wjones127) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:48]: > time spent in an ELT process is around converting data from one format to another

again, the schemas? Or the actual arrays / data?

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (wjones127) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:39]: It would be nice to have a diagram showing where the schemas fit in the flow of things. We have a `img` folder where diagrams can be added if need be.

**Review by wjones127 — CHANGES_REQUESTED** [review]
I think this could use some clarification to make sure readers of our blog understand what you have implemented more clearly.

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (yevgenypats) [reply _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md]: I can add flatbuffer. yes, we use both arrow arrays and types for data exchange.

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (yevgenypats) [reply _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:42]: Yes. I also pointed to our architecture docs for more info - https://www.cloudquery.io/docs/developers/architecture

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (yevgenypats) [reply _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:48]: both, we send arrow records over the wire

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (yevgenypats) [reply _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:39]: I pointed to our docs https://www.cloudquery.io/docs/developers/architecture but I can copy that img here as well

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (alamb) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:59]: A somewhat shameless self promotion is another advantage is that the mapping to / from the arrow type system and the parquet type system (including nested types) is supported by many of the arrow libraries as explained in 

https://arrow.apache.org/blog/2022/10/08/arrow-parquet-encoding-part-2/

**Review by alamb — APPROVED** [review]
I think this looks great -- thank you @yevgenypats  and @zeroshade

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (alamb) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:4]: Before this PR is merged, please update this date to the actual publish date and rename the file accordingly.

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (yevgenypats) [reply _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md]: Nice added!

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (yevgenypats) [reply _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md]: I don't have merge permissions but updated for today. mind merging it in ? :)

Comment (alamb) [reply _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md]: Hi @yevgenypats  -- thank you for the update. Given the discussion on the mailing list is only a few days old I recommend we wait a few more days to see if there are additional comments.

https://lists.apache.org/thread/mbbx6xzhp9zynkvncpw68j9f71p6gsbp

I can handle merging this (and updating the date) when it is ready, perhaps sometime mid week

File: _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md
Comment (wjones127) [inline _posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md:44]:
```suggestion
This is where a type system comes in. Source plugins extract information from APIs in the most performant way possible, defining a schema and then transforming the result from the API (JSON or any other format) to a well-defined type system. The destination plugin can then easily create the schema for its database and transform the incoming data to the destination types. So to recap, the source plugin sends mainly two things to a destination: 1) the schema 2) the records that fit the defined schema. In Arrow terminology, these are a schema and a record batch.
```

**Review by wjones127 — APPROVED** [review]
Thanks for the clarifications. I have one optional change but otherwise this looks good to me.

**Comment by alamb** [comment]
Unless anyone else would like to comment on this PR I plan to merge it tomorrow

**Comment by alamb** [comment]
FYI when I was testing this page locally I found that the cloud query blog link is broken

https://cloudquery.io/blog/adopting-apache-arrow-at-cloudquery
Screenshot 2023-05-04 at 6 31 06 AM

I am assuming this will go live after the arrow site; If that is not the case, we can make a new PR to fix the post post publish.

I will file a follow on PR to update the publish date. Thanks again everyone

**Comment by alamb** [comment]
The website is live: https://arrow.apache.org/blog/2023/05/04/adopting-apache-arrow-at-cloudquery/

**Comment by yevgenypats** [comment]
Thanks! Published on our end and backlinked to this post as well. Exciting stuff! https://www.cloudquery.io/blog/adopting-apache-arrow-at-cloudquery

**Comment by yevgenypats** [comment]
Also, tweeted - https://twitter.com/cloudqueryio/status/1654088502957555713 would love if you can either retweet or just publish on arrow twitter - it will help get some more visibility :)

**Pull request merged**
